pkg/trace/api/responses.go

```diff
@@ -90,18 +91,32 @@ func (wc *writeCounter) N() uint64 { return atomic.LoadUint64(&wc.n) }
 // httpRateByService outputs, as a JSON, the recommended sampling rates for all services.
 // It returns the number of bytes written and a boolean specifying whether the write was
 // successful.
-func httpRateByService(w http.ResponseWriter, dynConf *sampler.DynamicConfig) (n uint64, ok bool) {
+func httpRateByService(ratesVersion string, w http.ResponseWriter, dynConf *sampler.DynamicConfig) (n uint64, ok bool) {
+	wc := newWriteCounter(w)
+	var err error
+	defer func() {
+		n, ok = wc.N(), err == nil
+		if err != nil {
+			tags := []string{"error:response-error"}
+			metrics.Count(receiverErrorKey, 1, tags, 1)
+		}
+	}()
+
+	currentState := dynConf.RateByService.GetNewState(ratesVersion) // this is thread-safe
 	w.Header().Set("Content-Type", "application/json")
+	w.Header().Set(headerRatesPayloadVersion, currentState.Version)
+
+	if ratesVersion != "" && ratesVersion == currentState.Version {
+		// rates weren't changed
+		_, err = wc.Write([]byte("{}"))
+		return
+	}
+
 	response := traceResponse{
-		Rates: dynConf.RateByService.GetAll(), // this is thread-safe
+		Rates:      currentState.Rates,
+		Mechanisms: currentState.Mechanisms,
```

Comment:
In this implementation, we return `Mechanisms` even `Version` header doesn't exist. Which means, older tracers might receive extra field (`mechanism`) in the API response.
I did like this, because I think the tracers will safely ignore fields they don't recognize.

Reply:
We decided not to return `mechanisms` when `Version` doesn't exist. I'll update the PR and RFC, accordingly.